File: 0735_asteroid_collision/asteroid_collision.py
from typing import List


def asteroid_collision(asteroids: List[int]) -> List[int]:
    """
    Time: O(n)
    Space: O(n)
    """
    res = []
    for a in asteroids:
        while len(res) and (a < 0 and res[-1] > 0):
            if -a == res[-1]:
                res.pop()
                break
            elif -a > res[-1]:
                res.pop()
                continue
            else:
                # breaking out of while-else does not result in else
                break
        else:
            res.append(a)
    return res

File: 0735_asteroid_collision/asteroid_collision.java
import java.util.LinkedList;

class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        LinkedList<Integer> res = new LinkedList<>();
        for (int a : asteroids) {
            if (a > 0) {
                res.add(a);
            } else {
                while (!res.isEmpty() && res.getLast() > 0 && res.getLast() < -a) {
                    res.pollLast();
                }
                if (!res.isEmpty() && res.getLast() == -a) {
                    res.pollLast();
                } else if (res.isEmpty() || res.getLast() < 0) {
                    res.add(a);
                }
            }
        }
        return res.stream().mapToInt(i -> i).toArray();
    }
}